File: week4/java/Permutations_46.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        permute(nums, new ArrayList<Integer>(), result);
        return result;
    }
    
    private void permute(int[] nums, List<Integer> list, List<List<Integer>> result) {
        if (list.size() == nums.length) {
            result.add(new ArrayList<Integer>(list));
            return;
        }
        for (int num : nums) {
            if (list.contains(num)) continue;
            list.add(num);
            permute(nums, list, result);
            list.remove(list.size() - 1);
        }
    }
}

File: week4/python/Permutations_46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        result = list()
        
        def permutation(pemur: List[int], visited: List[int]):
            if len(pemur) == len(nums):
                result.append(list(pemur))
                return
            
            for i in range(len(nums)):
                num = nums[i]
                if not visited[i]:
                    visited[i] = True
                    pemur.append(num)
                    permutation(pemur, visited)
                    visited[i] = False
                    pemur[:] = pemur[:-1]
        

        permutation(list(), [False] * len(nums))
        return result
